File: question/leetcode/leetcode0141.Linked-List-Cycle/leetcode0141.go
package leetcode

import "clarencedu/algorithm-go/structure"

type ListNode = structure.ListNode

func hasCycle(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return false
	}
	slow, fast := head, head

	for slow != fast {
		if fast == nil || fast.Next == nil {
			return false
		}
		slow = slow.Next
		fast = fast.Next.Next
	}
	return true
}

File: question/leetcode/leetcode0141.Linked-List-Cycle/leetcode0141.ts
// class ListNode {
// 	val: number
// 	next: ListNode | null
// 	constructor(val?: number, next?: ListNode | null) {
// 		this.val = (val === undefined ? 0 : val)
// 		this.next = (next === undefined ? null : next)
// 	}
// }

function hasCycle(head: ListNode | null): boolean {
	if (head === null) {
		return false;
	}
	let slow = head, fast = head;
	let hasCircle: boolean = false;
	while (fast !== null && fast.next !== null) {
		slow = slow.next!;
		fast = fast.next.next!;
		if (slow === fast) {
			hasCircle = true;
			break;
		}
	}

	return hasCircle
};